### PR TITLE
Add Support for US Government cloud deployment

### LINKED ISSUE #3128: Enable deployment to Azure Government Cloud
Deployment fails immediately when attempting to deploy to Azure Gov regions because of the distinct endpoints hardcoded into TRE.  Requesting assisting in removing hardcoded endpoints and allowing endpoints to be determined based on region.  Anywhere the az libraries are used (so long as it is up to date) will use correct endpoints.

### PR DESCRIPTION
**Pull request by LizaShak**
# Resolves #3128 

## What is being addressed

Enable deployment on US Government cloud. 

## How is this addressed

- Use a separate module to get privatelinks
- Remove hardcoded endpoints
- Pass additional parameters on clients
- Specify parameters needed to determine the destination cloud

### CODE REVIEW

File: airlock_processor/shared_code/cloud.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    try:
+        arm_environment = os.environ["ARM_ENVIRONMENT"].lower()
+    except KeyError as e:
+        logging.error(f'Missing environment variable: {e}')
+        raise
+    return arm_environment
+
+
+# Get active cloud information such as endpoints and suffixes
+def get_cloud() -> cloud.Cloud:
```
[the remark this answers is not in the document]
Comment (martinpeck) [reply airlock_processor/shared_code/cloud.py:16]: i.e. repeated in api_app/core/cloud.py

File: cli/tre/authentication.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        authority=AuthorityBuilder(instance=get_aad_authority_fqdn(), tenant=aad_tenant_id),
+        token_cache=token_cache)
+
+
+def get_cloud() -> cloud.Cloud:
+    return cloud.get_active_cloud()
+
+
+def get_aad_authority_fqdn() -> str:
+    return get_cloud().endpoints.active_directory.replace('https://', '')
```
Comment (martinpeck) [inline cli/tre/authentication.py:48]: nit: it's generally better to use `urllib` to parse and manipulate URLs

Comment (martinpeck) [reply cli/tre/authentication.py:48]: also, this is repeated code
perhaps we need a TRE utils package that we can import into multiple things

File: core/terraform/locals.tf [hunk 3]
```diff
@@ -28,8 +31,9 @@ locals {
     "privatelink.cert.api.azureml.ms",
     "privatelink.notebooks.azure.net",
     "privatelink.postgres.database.azure.com",
-    "nexus-${var.tre_id}.${var.location}.cloudapp.azure.com",
     "privatelink.mysql.database.azure.com",
     "privatelink.azuredatabricks.net"
   ])
+
+  service_bus_namespace_fqdn = regex("(?:(?P<scheme>[^:/?#]+):)?(?://(?P<fqdn>[^/?#:]*))?(?::(?P<port>[0-9]+))?(?P<path>[^?#]*)(?:\\?(?P<query>[^#]*))?(?:#(?P<fragment>.*))?", azurerm_servicebus_namespace.sb.endpoint).fqdn
```
Comment (martinpeck) [inline core/terraform/locals.tf:38]: image

Comment (martinpeck) [reply core/terraform/locals.tf:38]: is there any way we can pad this with comments so it's more obvious what it's expecting, the things it captures, and what it outputs?

File: templates/shared_services/airlock_notifier/terraform/locals.tf [hunk 4]
```diff
@@ -9,5 +9,5 @@ locals {
     tre_id                = var.tre_id
     tre_shared_service_id = var.tre_resource_id
   }
-  default_tre_url = "https://${var.tre_id}.${data.azurerm_resource_group.core.location}.cloudapp.azure.com"
+  default_tre_url = "https://${data.azurerm_public_ip.app_gateway_ip.fqdn}"
```
Comment (martinpeck) [inline templates/shared_services/airlock_notifier/terraform/locals.tf:12]: How does the TRE ID get added?
Or...does it?

Comment (martinpeck) [reply templates/shared_services/airlock_notifier/terraform/locals.tf:12]: Ah....we're using the FQDN, that already has it, right?

File: templates/shared_services/cyclecloud/terraform/variables.tf [hunk 5]
```diff
@@ -1,2 +1,3 @@
 variable "tre_id" {}
 variable "tre_resource_id" {}
+variable "arm_environment" {}
```
Comment (martinpeck) [inline templates/shared_services/cyclecloud/terraform/variables.tf:3]: nit: add new-line to end of file

File: templates/shared_services/databricks-auth/terraform/variables.tf [hunk 6]
```diff
@@ -7,3 +7,5 @@ variable "tre_resource_id" {
   type        = string
   description = "Unique TRE Resource ID"
 }
+
+variable "arm_environment" {}
```
Comment (martinpeck) [inline templates/shared_services/databricks-auth/terraform/variables.tf:11]: nit: add new-line to end of file

File: templates/shared_services/gitea/terraform/variables.tf [hunk 7]
```diff
@@ -29,3 +29,5 @@ variable "acr_name" {
   type        = string
   description = "Name of Azure Container Registry"
 }
+
+variable "arm_environment" {}
```
Comment (martinpeck) [inline templates/shared_services/gitea/terraform/variables.tf:33]: nit: add new-line to end of file

File: templates/workspace_services/databricks/terraform/variables.tf [hunk 8]
```diff
@@ -22,3 +22,5 @@ variable "is_exposed_externally" {
   type        = bool
   description = "If the databricks workspace is exposed externally or not."
 }
+
+variable "arm_environment" {}
```
Comment (martinpeck) [inline templates/workspace_services/databricks/terraform/variables.tf:26]: nit: add new-line to end of file

File: templates/workspace_services/health-services/terraform/variables.tf [hunk 9]
```diff
@@ -42,3 +47,5 @@ variable "auth_client_secret" {
   type        = string
   description = "Used to authenticate into the AAD Tenant to get app role members"
 }
+
+variable "arm_environment" {}
```
Comment (martinpeck) [inline templates/workspace_services/health-services/terraform/variables.tf:51]: nit: add new-line to end of file

File: templates/workspace_services/innereye/terraform/variables.tf [hunk 10]
```diff
@@ -9,3 +9,4 @@ variable "arm_use_msi" {
 }
 variable "inference_sp_client_id" {}
 variable "inference_sp_client_secret" {}
+variable "arm_environment" {}
```
Comment (martinpeck) [inline templates/workspace_services/innereye/terraform/variables.tf:12]: nit: add new-line to end of file

File: templates/workspace_services/mlflow/terraform/variables.tf [hunk 11]
```diff
@@ -10,3 +10,4 @@ variable "is_exposed_externally" {
   description = "Is the webapp available on the public internet"
   default     = false
 }
+variable "arm_environment" {}
```
Comment (martinpeck) [inline templates/workspace_services/mlflow/terraform/variables.tf:13]: nit: add new-line to end of file

File: templates/workspace_services/mysql/terraform/variables.tf [hunk 12]
```diff
@@ -10,3 +10,4 @@ variable "storage_mb" {
     error_message = "The storage value is out of range."
   }
 }
+variable "arm_environment" {}
```
Comment (martinpeck) [inline templates/workspace_services/mysql/terraform/variables.tf:13]: nit: add new-line to end of file

**Review by martinpeck — APPROVED** [review]
Wow. That was epic! Well done @LizaShak !

I've made a few comments, but they're mostly minor things.
Please ensure you get reviews from others (don't merge purely based upon my approval).

File: airlock_processor/shared_code/cloud.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (martinpeck) [reply airlock_processor/shared_code/cloud.py:16]: Repeated code is a repo-wide issue

If you think this specific case is fixable in this PR then great, but feel free to resolve this as "won't fix". It's fine as-is.

File: cli/tre/authentication.py (hunk 2 above)
Comment (martinpeck) [reply cli/tre/authentication.py:48]: Repeated code is a repo-wide issue

If you think this specific case is fixable in this PR then great, but feel free to resolve this as "won't fix". It's fine as-is.

File: core/terraform/locals.tf (hunk 3 above)
Comment (martinpeck) [reply core/terraform/locals.tf:38]: This is really the only conversation I have open that I think needs addressing.
If we can, I'd suggest some lines of comment above this regex to explain what it does and how it works, and what the expected inputs/outputs are.

**Review by tamirkamara — APPROVED** [review]
Great job!

**Pull request merged**